libcontainer/seccomp/seccomp_linux.go

```diff
@@ -15,67 +15,90 @@ import (
 )
 
 var (
-	actAllow = libseccomp.ActAllow
-	actTrap  = libseccomp.ActTrap
-	actKill  = libseccomp.ActKill
-	actTrace = libseccomp.ActTrace.SetReturnCode(int16(unix.EPERM))
-	actLog   = libseccomp.ActLog
-	actErrno = libseccomp.ActErrno.SetReturnCode(int16(unix.EPERM))
+	actAllow  = libseccomp.ActAllow
+	actTrap   = libseccomp.ActTrap
+	actKill   = libseccomp.ActKill
+	actTrace  = libseccomp.ActTrace.SetReturnCode(int16(unix.EPERM))
+	actLog    = libseccomp.ActLog
+	actErrno  = libseccomp.ActErrno.SetReturnCode(int16(unix.EPERM))
+	actNotify = libseccomp.ActNotify
 )
 
 const (
 	// Linux system calls can have at most 6 arguments
 	syscallMaxArguments int = 6
 )
 
-// Filters given syscalls in a container, preventing them from being used
-// Started in the container init process, and carried over to all child processes
-// Setns calls, however, require a separate invocation, as they are not children
-// of the init until they join the namespace
-func InitSeccomp(config *configs.Seccomp) error {
+// InitSeccomp installs the seccomp filters to be used in the container as
+// specified in config.
+// Returns the seccomp file descriptor if any of the filters include a
+// SCMP_ACT_NOTIFY action, otherwise returns -1.
+func InitSeccomp(config *configs.Seccomp) (int, error) {
 	if config == nil {
-		return errors.New("cannot initialize Seccomp - nil config passed")
+		return -1, errors.New("cannot initialize Seccomp - nil config passed")
 	}
 
 	defaultAction, err := getAction(config.DefaultAction, config.DefaultErrnoRet)
 	if err != nil {
-		return errors.New("error initializing seccomp - invalid default action")
+		return -1, errors.New("error initializing seccomp - invalid default action")
+	}
+
+	// Ignore the error since pre-2.4 libseccomp is treated as API level 0.
+	apiLevel, _ := libseccomp.GetAPI()
+	for _, call := range config.Syscalls {
+		if call.Action == configs.Notify {
+			if apiLevel < 6 {
+				return -1, fmt.Errorf("seccomp notify unsupported: API level: got %d, want at least 6. Please try with libseccomp >= 2.5.0 and Linux >= 5.7", apiLevel)
+			}
+
+			if call.Name == "write" {
+				return -1, errors.New("SCMP_ACT_NOTIFY cannot be used for the write syscall")
+			}
```

Comment:
Is there a reason for this restriction? I can imagine how it might be unsafe (or unwise) to do this on paper, but is it really necessary to disallow this within runc?

Reply:
Yes, basically it will hang for ever if we allow it. We need the write syscall to send the seccomp fd over the pipe to the parent (we write to the pipe [here][link]), then the parent gets the fd with pifd_getfd and sends it over the unix socket to the seccomp agent.

If the write syscall has the notify action, we can't never send a msg to the parent, which will mean the seccomp agent won't get the seccomp fd and the write will be blocked forever.

However read/close (functions that we also use to sync with parent via the pipe) are not restricted as it is possible to send the seccomp fd to the parent, the seccomp agent gets it and the first few notifications it will get if read/close are handled are for the ones done in `syncParentSeccomp()` (and quite trivially allow to proceed syscalls on fds that are not known by the agent, for example).

I will add a comment to explain this in the code, if that sounds good to you @cyphar 

[link]: https://github.com/opencontainers/runc/pull/2682/commits/10ad86bd09910557e2c3680c97971fbfd7b1c60d#diff-84fffb149f8b86c903e09613cbebad12de596f4139d54e288388b1458e08b8ecR288

Reply:
@cyphar Added a comment explaining this [here][link]


[link]: https://github.com/opencontainers/runc/pull/2682/files#diff-3693bdb9f47092da6d6e35007a1689f74ddc39be8e017d475a593481c437ac3bR55-R73

Reply:
Ah of course. Yeah that's a bit sad :/ -- but we can always figure out some workaround for this in the future. (I wonder how the LXC folks worked around this problem -- though I think they don't have a similar write-based synchronisation model since they're written in C which means you don't need to do so much busy work.)

Reply:
@cyphar I've asked Christian Brauner [in the past](https://lore.kernel.org/lkml/20210301110907.2qoxmiy55gpkgwnq@wittgenstein/). It seems they use clone3 sharing the fds table and use a fixed number for the seccomp notify fd. Then, they get it with pidfd_getfd().

Not sure how LXC uses a fixed fd number for seccomp notify, though.

Reply:
You could just pick an arbitrarily high fd and use `dup3` to clone it to the right fd number. It might be interesting to look into whether we could do it with `pidfd_getfd` though since we'd need to have a fallback for quite a while, it'd probably complicate the code a bit too much...

Reply:
Oh, right, that was easy. Yeah, for runc I think is simpler to use the pipe, that is the main sync mechanism used tree-wide. We are using `pidfd_getfd` here, though: we sync with the parent to let them know when the fd is ready and send the fd plain number, then the parent gets it with `pidfd_getfd`

Reply:
Ah, I forgot that we are requiring Linux 5.7 or newer for the feature -- pidfd_getfd isn't available for the earlier kernels with seccomp notify support.

Reply:
And 5.7 is also enforced by libseccomp-golang, that only supports notification with tsync (requires libseccomp api level 6). This is as the vast majority of go programs will need tsync too. [Here](https://github.com/seccomp/libseccomp-golang/pull/59#discussion_r615949270) is the conversation we had with upstream about this when writing the patches for libseccomp-golang

Before 5.7, notify and tsync where mutually exclusive (torvalds/linux@7a0df7f). From Linux 5.7 introduced a new API to enable seccomp notification with tsync together in this commit torvalds/linux@5189149.